### PR TITLE
Flow decomposition observers refactoring

### PR DESCRIPTION
**Pull request by caioluke**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

Currently, while running a flow decomposition, we can only retrieve intermediate results for branches and nodes. We can't retrieve an intermediate state for a boundary node, for example.

**What is the new behavior (if this is a feature change)?**
Flow decomposition observers now have full access to the network after an AC LF and a DC LF. Therefore, the user can retrieve any information they need from the network at those states.

**Does this PR introduce a breaking change or deprecate an API?**
- [X] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [X] The *Breaking Change* or *Deprecated* label has been added
- [x] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

Implement the new methods from the FlowDecompositionObserver interface
- computedAcLoadFlowResults
- computedDcLoadFlowResults)

As well as remove the old ones
- computedAcNodalInjections
- computedDcNodalInjections
- computedAcFlowsTerminal1
- computedAcFlowsTerminal2
- computedDcFlows
- computedAcCurrentsTerminal1
- computedAcCurrentsTerminal2)

### CODE REVIEW

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java [hunk 1]
```diff
@@ -74,52 +76,17 @@ public interface FlowDecompositionObserver {
     void computedPsdfMatrix(Map<String, Map<String, Double>> psdfMatrix);
 
     /**
-     * Called when the AC nodal injections matrix is computed (for base case or contingency)
+     * Called after an AC loadflow is launched and terminated
```
Comment (OpenSuze) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java:79]:
```suggestion
     * Called after an AC loadflow has been computed
```

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java [hunk 2]
```diff
@@ -74,52 +76,17 @@ public interface FlowDecompositionObserver {
...
-     * @param positions the positions after AC loadflow
+     * @param network the network after AC loadflow
      * @param fallbackHasBeenActivated true if AC loadflow didn't converge
      */
-    void computedAcNodalInjections(Map<String, Double> positions, boolean fallbackHasBeenActivated);
+    void computedAcLoadFlowResults(Network network, boolean fallbackHasBeenActivated);
 
     /**
-     * Called when the DC nodal injections matrix is computed (for base case or contingency)
+     * Called after a DC loadflow is launched and terminated
```
Comment (OpenSuze) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java:87]:
```suggestion
     * Called after a DC loadflow has been computed
```

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java [hunk 3]
```diff
@@ -74,52 +76,17 @@ public interface FlowDecompositionObserver {
...
     /**
-     * Called when the AC nodal injections matrix is computed (for base case or contingency)
+     * Called after an AC loadflow is launched and terminated
      *
-     * @param positions the positions after AC loadflow
+     * @param network the network after AC loadflow
      * @param fallbackHasBeenActivated true if AC loadflow didn't converge
      */
-    void computedAcNodalInjections(Map<String, Double> positions, boolean fallbackHasBeenActivated);
+    void computedAcLoadFlowResults(Network network, boolean fallbackHasBeenActivated);
```
Comment (OpenSuze) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java:84]: Could you add an parameter "LoadFlowResult" ?
```suggestion
    void computedAcLoadFlowResults(Network network, LoadFlowResult loadFlowResult, boolean fallbackHasBeenActivated);
```

File: flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java [hunk 4]
```diff
@@ -74,52 +76,17 @@ public interface FlowDecompositionObserver {
...
-     */
-    void computedAcCurrentsTerminal1(Map<String, Double> currents);
-
-    /**
-     * Called when the AC loadflow is computed (for base case or contingency)
-     *
-     * @param currents the terminal 1 current for all branches
-     */
-    void computedAcCurrentsTerminal2(Map<String, Double> currents);
+    void computedDcLoadFlowResults(Network network);
```
Comment (OpenSuze) [inline flow-decomposition/src/main/java/com/powsybl/flow_decomposition/FlowDecompositionObserver.java:91]: Could you add an parameter "LoadFlowResult" please ?

File: docs/flow_decomposition/flow-decomposition-detailed-results.md
Comment (OpenSuze) [inline docs/flow_decomposition/flow-decomposition-detailed-results.md:1]: Could you update this documentation ?

**Review by OpenSuze — APPROVED** [review]
Thank you for this PR !

File: flow-decomposition/src/test/java/com/powsybl/flow_decomposition/FlowDecompositionObserverTest.java
Comment (OpenSuze) [inline flow-decomposition/src/test/java/com/powsybl/flow_decomposition/FlowDecompositionObserverTest.java:1]: Could you test the `loadFlowResult` ? At least that they are not null ?

Comment (caioluke) [reply flow-decomposition/src/test/java/com/powsybl/flow_decomposition/FlowDecompositionObserverTest.java:1]: Done!

**Review by OpenSuze — APPROVED** [review]
Thank you for this PR, I approve this !

**Pull request merged**